aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -117,6 +118,39 @@ Tensor pinverse(const Tensor& self, double rcond) {
   return V.mm(S_pseudoinv.diag().mm(U.t()));
 }
 
+Tensor _matrix_rank_helper(const Tensor& self, bool symmetric) {
+  Tensor S;
+  if (!symmetric) {
+    Tensor U, V;
+    std::tie(U, S, V) = self.svd();
+  } else {
+    Tensor eigvecs;
+    std::tie(S, eigvecs) = self.symeig();
+    S = S.abs();
+  }
+  return S;
+}
+
+Tensor matrix_rank(const Tensor& self, double tol, bool symmetric) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()) && self.dim() == 2,
+           "matrix_rank(", self.type(), "{", self.sizes(), "}): expected a 2D tensor "
+           "of floating types");
+
+  Tensor S = _matrix_rank_helper(self, symmetric);
+  return (S > tol).sum();
+}
+
+Tensor matrix_rank(const Tensor& self, bool symmetric) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()) && self.dim() == 2,
+           "matrix_rank(", self.type(), "{", self.sizes(), "}): expected a 2D tensor "
+           "of floating types");
+
+  Tensor S = _matrix_rank_helper(self, symmetric);
+  Tensor tol = S.max() * _get_epsilon(self.type());
```

Comment:
Thanks! 

Suggestions: 
1. calculate `_get_epsilon(self.type()) * std::max(self.size(0), self.size(1));` first for one less kernel launch.
2. we can use implace `mul_` here.

Reply:
Btw, can you also remove the `toDouble()` on `pinv` while you are at this? Thanks!